R/cor_test_biserial.R

```diff
@@ -0,0 +1,73 @@
+#' @keywords internal
+.cor_test_biserial <- function(data, x, y, ci = 0.95, method = "biserial", ...) {
+
+  # Sanity
+  if (.vartype(data[[x]])$is_binary & .vartype(data[[y]])$is_binary == FALSE){
+    binary <- x
+    continuous <- y
+  } else if (.vartype(data[[y]])$is_binary & .vartype(data[[x]])$is_binary == FALSE){
+    binary <- y
+    continuous <- x
+  } else{
+    stop("Biserial and point-biserial correlations can only be applied for one dichotomous and one continuous variables.")
+  }
+
+
+  # Get biserial correlation
+  if(method == "biserial"){
+    out <- .cor_test_biserial_biserial(data, x, y, continuous, binary, ci)
+
+  # Get point-biserial correlation
+  } else{
+    out <- .cor_test_freq(data, continuous, binary, ci = ci, method = "pearson", ...)
+    out$Parameter1 <- x
+    out$Parameter2 <- y
+    out$Method <- "Point-biserial"
+  }
+
+  out
+}
+
+
+
+
+
+
+
+
+
+
+
+#' @keywords internal
+.cor_test_biserial_biserial <- function(data, x, y, continuous, binary, ci){
+
+  # TODO: get rid off psych https://www.statisticshowto.datasciencecentral.com/point-biserial-correlation/
```

Comment:
@strengejacke @mattansb the maths geniuses there's an easy formula https://www.statisticshowto.datasciencecentral.com/point-biserial-correlation/ for the biserial correlation so that we don't depend on psych... care to have a look? 😁

Reply:
Maybe like:

``` r
set.seed(123)
y <- rbinom(100, 1, .3)
x <- rnorm(100)

m1 <- mean(x[y == 1])
m0 <- mean(x[y == 0])
sn <- sd(x)
q <- mean(y)
p <- 1 - q

((m1 - m0) / sn) * sqrt(p * q)
#> [1] 0.06151908


y2 <- y
y2[y == 0] <- "a"
y2[y == 1] <- "f"

y3 <- performance:::.factor_to_numeric(y2, lowest = 0)

m1 <- mean(x[y == 1])
m0 <- mean(x[y == 0])
sn <- sd(x)
q <- mean(y)
p <- 1 - q

((m1 - m0) / sn) * sqrt(p * q)
#> [1] 0.06151908
```

<sup>Created on 2020-03-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

Reply:
ups, that was point biseral...

Reply:
Here you are:

``` r
own_biserial <- function(x, y) {
  cc <- complete.cases(x, y)
  x <- x[cc]
  y <- y[cc]
  
  y <- performance:::.factor_to_numeric(y, lowest = 0)
  
  m1 <- mean(x[y == 1])
  m0 <- mean(x[y == 0])
  sn <- sd(x)
  q <- mean(y)
  p <- 1 - q
  
  zp <- dnorm(qnorm(q))
  
  (((m1 - m0) * (p * q / zp)) / sd(x))
}

set.seed(123)
y <- rbinom(100, 1, .3)
x <- rnorm(100)

own_biserial(x, y)
#> [1] 0.08155037
psych::biserial(x, y)
#>            [,1]
#> [1,] 0.08155037


set.seed(456)
y <- rbinom(100, 1, .3)
x <- rnorm(100)

own_biserial(x, y)
#> [1] 0.02964972
psych::biserial(x, y)
#>            [,1]
#> [1,] 0.02964972
```

<sup>Created on 2020-03-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>